TAP-2.md

```diff
@@ -37,123 +37,84 @@ Provide all on-chain liquidity in a familiar limit orderbook interface to read (
 ### Nice-to-have Requirements
 - **Local fine-tuning**: Traders can request not only the provided points on the orderbook but also any arbitrary point in between. Those calls run a iterative search (solver) for the correct amount – within a certain window of tolerance.
 ## Not included
-# Implementation
+# Design (draft)
 *Just a draft for inspiration – take and leave what you like.*
-```ts
-// Core Domain Types
-type Side = 'buy' | 'sell'
-type GasMode = 'include' | 'exclude'
+## Core Components
+### 1. Protocol Stream Manager
+Manages connection to Tycho Indexer for real-time protocol state updates.
+```rust
+use tycho_core::dto::Chain;
+use tycho_client::stream::TychoStreamBuilder;
 
-type TokenPair = {
-  token0: string  // Base token address
-  token1: string  // Quote token address
-}
-
-type Route = {
-  protocols: string[]  // DEX protocols used
-  calldata: string    // Executable calldata including splits
-}
-
-type Level = {
-  price: string
-  size: string
-  route: Route
-  gasCost: {
-    wei: string
-    token0: string
-    token1: string
-  }
-}
-
-// Book State
-type OrderbookState = {
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
-  bids: Level[]  // Sorted best to worst
-  asks: Level[]  // Sorted best to worst
-}
-
-// Query Options
-type OrderbookOptions = {
-  maxHops?: number
-  maxSplits?: number
-  includedProtocols?: string[]
-  pricePoints?: number
-  gasMode: GasMode
-  gasToken?: string
-  maxGasPriceGwei?: string
-}
-
-type FinetuneOptions = OrderbookOptions & {
-  tolerance: string
-  maxIterations?: number
+pub struct ProtocolStreamManager {
+    tycho_url: String,
+    auth_key: Option<String>,
+    tvl_threshold: f64,
 }
+```
+### 2. Router Interface
+An interface for routing solutions. Implement a simple solver, and expect to later integrate a more sophisticated external solver.
+```rust
+/// Generic router interface for finding and executing trades
+pub trait Router {
+    fn get_amount_out(
+        &self,
+        amount_in: U256,
+        token_in: &Address,
+        token_out: &Address,
+    ) -> Result<GetAmountOutResult, Error>;
 
-// Operation Results
-type SizeQuote = {
-  effectivePrice: string
-  route: Route
-  gasCost: Level['gasCost']
-  priceImpact: string
+    fn encode_solution(
+        &self,
+        result: GetAmountOutResult,
+        sender: Address,
+        receiver: Address,
+    ) -> Result<RoutingSolution, Error>;
```

Comment:
Not sure what you want to achieve with this interface? It seems like it should be internal and not public.. or is this something a user of the tycho-orderbook library should have access to? If it is internal I would not mention it here.

Reply:
Ok coming back to this after reading the specs below. You want to allow the router implementation to be replaced later on. In this case it may make sense specifying this trait, but I would instead just mention this as a requirement and let the developer specify the interface. If you want to keep this as an example it is ok, but probably it would make sense to accept a batch of `get_amount_out` requests here especially if we plan to delegate this to any external router APIs.

Reply:
Its internal – I'm trying to help the engineer building this project understand what components he needs to build - not define an interface for users.

Reply:
> accept a batch of get_amount_out requests here especially if we plan to delegate this to any external router APIs.

Everything is meant to stay local – i.e. the "external" router will be another local rust solver.